### PR TITLE
Fix interstitial sprout anchors check

### LINKED ISSUE #3210: Document consensus rules from Zcash spec: Section 3.5 JoinSplit Transfers and Descriptions
For each consensus rule:

- check that the rule is implemented in Zebra
- make sure the latest version of the rule is quoted in the docs for the code that implements it
- make sure the code implements the latest version of the rule

## Consensus rules

https://zips.z.cash/protocol/protocol.pdf#joinsplit
Screenshot 2021-12-17 at 13 44 37

### LINKED ISSUE #3236: Fix Sprout anchors check
## Motivation

There are at least two blocks with transactions that fail to validate the Sprout anchor check.

[We disabled the check for now](https://github.com/ZcashFoundation/zebra/pull/3206), but we must investigate the issue, fix it, and restore the check.

One of the transactions is https://explorer.zcha.in/transactions/5e53ebca23efd85774677f342e8dca5493865b3892dcf344c94284fa50d11102, the fifth tx in block `1_047_908`, which fails to find the anchor `019c435cd1e8aca9a4165f7e126ac6e548952439d50213f4d15c546df9d49b61` for its second JoinSplit.

Another missing anchor is `3ad623811ffa4fe8498b23f3d6bb4e086dca32269afef6c8e572fd9ee6d0c0ea` which seems to be used in a block near `1_057_736`.

## Specifications & Solutions

We currently fetch the output note commitment tree from the most recently validated block, and we build the interstitial note commitment tree by adding note commitments from the JoinSplits in the tx being currently validated by following the linear order of the JoinSplits in the tx. This turns out to be incorrect.
 
The issue is that we should look up and fetch the tree that the anchor of the previous JoinSplit refers to, and add the commitments of the previous JoinSplit to this tree. The root of this new tree is available for a check against the anchor of the current JoinSplit. This new tree is also available for the next JoinSplit. (Since the anchor of the current JoinSplit might refer to it, and the current JoinSplit will become the previous one for the next JoinSplit).
 
Spec:
image

For a reference, here's a transaction builder: https://github.com/zcash/zcash/blob/2cee089697ef817ca5c3d26d0c5db973e009037f/src/transaction_builder.cpp#L622. The relevant code is between lines 622 and 655.

The solution will require fetching a note commitment tree for each Sprout anchor. Note that multiple anchors might refer to the same tree since the roots of non-interstitial note commitment trees are stored on a per-block basis. We discussed that storing the trees only for the anchors in the non-finalized state might be sufficient. We don't need to store multiple trees for Sapling or Orchard because these shielded pools don't require fetching the trees according to anchors.

Consensus rules:
image

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

There was an issue with the interstitial Sprout anchor check, described in https://github.com/ZcashFoundation/zebra/issues/3236

### Specifications

image

https://zips.z.cash/protocol/protocol.pdf#joinsplit

## Solution

- Change state to add the matching note commitment tree with each Sprout anchor
- Also add it to non-finalized state (`Chain`)
- When checking the anchor, build the interstitial trees starting from the tree that matches the anchor (and not from the tree of the finalized tip)
- Update state documentation

Closes #3236 and #3210.

## Review

Not particularly urgent, but it's the final consensus check. Usually @dconnolly and/or @upbqdn could review but @teor2345 might want to take a look.

After accepting but before merging we need to update the cached state in CI (or disable the CI job until we do that).

There is no specific test - I tested by syncing to the tip:

image

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: book/src/dev/rfcs/0005-state-updates.md
Comment (upbqdn) [inline book/src/dev/rfcs/0005-state-updates.md]:
```suggestion
  regardless of where they come from. The exception is `sprout_anchors` which also maps
```

File: zebra-state/src/service/check/anchors.rs
Comment (upbqdn) [inline zebra-state/src/service/check/anchors.rs]: `interstitial_trees` might be a better name?

**Review by upbqdn — DISMISSED** [review]
All looks good. It's handy that the commitment trees could be stored under `sprout_anchors`.
I left only a few minor suggestions.

**Comment by upbqdn** [comment]
I can regenerate the cached state once someone else approves this PR.

File: zebra-state/src/service/check/anchors.rs
Comment (upbqdn) [inline zebra-state/src/service/check/anchors.rs]: The order of the queries seems like the most optimal one. 👍

**Review by teor2345 — COMMENTED** [review]
Looks good to me, but it needs a cached state rebuild.

**Comment by teor2345** [comment]
@upbqdn Conrado isn't available this week, so can you push all the changes you think are needed to this PR's branch?

I can re-approve after you're done.

> I can regenerate the cached state once someone else approves this PR.

I think it's ready to go, after you make whatever changes you think are needed.

Just checking, are you going to regenerate the cached state, update the image in this PR, then merge it to `main`?
(That way, the tests won't fail on `main`, because there's no gap between the version and cached state update.)

The hash in this line needs to be updated:
https://github.com/ZcashFoundation/zebra/blob/454897bc330633246ab2ba714a17c97d8c823bc5/.github/workflows/test.yml#L57

**Comment by upbqdn** [comment]
> Just checking, are you going to regenerate the cached state, update the image in this PR, then merge it to main?

Yes, exactly. I'm going to follow the steps in the onboarding doc. I refined them recently, so it should be smooth.

File: zebra-state/src/service/check/anchors.rs
Comment (daira) [inline zebra-state/src/service/check/anchors.rs]: I would include the full quote here; it's not too much longer. (The missing part is "either some earlier block's final Sprout treestate, or", which is important.)

**Review by daira — COMMENTED** [review]
Looks good. I apologize again for the hairiness of the spec here.

File: zebra-state/src/service/check/anchors.rs
Comment (upbqdn) [reply zebra-state/src/service/check/anchors.rs]: Thank you @daira. I added that, and I rewrote the docs so that they comply with the new way of documenting the consensus rules. We recently agreed on a new way of documenting the consensus rules to facilitate the upcoming audit.

File: zebra-state/src/service/check/anchors.rs
Comment (upbqdn) [inline zebra-state/src/service/check/anchors.rs:118]: This consensus rule is naturally implemented in different parts of the codebase. The implementation below effectively deals only with the first sentence of the consensus rule. What's the best way to deal with a situation like this?

I can think of either splitting the consensus rule and mentioning only the relevant part or quoting the whole consensus rule in multiple places in the codebase. None of that seems ideal.

Comment (upbqdn) [inline zebra-state/src/service/check/anchors.rs]: This consensus rule is naturally implemented in different parts of the codebase. The implementation below effectively deals only with the first sentence of the consensus rule. What's the best way to deal with a situation like this?

I can think of either splitting the consensus rule and mentioning only the relevant part or quoting the whole consensus rule in multiple places in the codebase. None of that seems ideal.

Comment (teor2345) [reply zebra-state/src/service/check/anchors.rs]: We're trying to document consensus rules where they are enforced in the code. Because those locations use the types or functions that implement the rules. So that should limit what we need to document.

But if a rule is enforced in multiple places, I suggest:
1. quote the full rule in both places
2. document the split in both places, with something like:

```ascii
/// This rule is also implemented by [`(type or module link)`].
/// The "(partial rule summary)" check is implemented here.
```

But we have a lot of flexibility here, so let's see how it looks for this specific rule?

Once we have a format we like, let's update the pinned template in Discord. (Or we could put a short consensus rule doc template in `zebra/book/src/dev`).

Comment (upbqdn) [reply zebra-state/src/service/check/anchors.rs]: OK, I'll do it as you suggested.

**Comment by upbqdn** [comment]
The last bit that's missing here is the new state. I will generate it once we fix the other issues.

**Comment by teor2345** [comment]
> The last bit that's missing here is the new state. I will generate it once we fix the other issues.

Can you review PR https://github.com/ZcashFoundation/zebra/pull/3168 ?

It might fix the hangs by spawning the syncer as a separate task.

Or we might also need to fix #3234 and #3254.

**Comment by upbqdn** [comment]
This PR also partially documents the consensus rules for #3211.

**Comment by upbqdn** [comment]
The following statement should not be a consensus rule
image
 
 This PR documents all consensus rules mentioned in #3210 except for the statement above, so it also closes #3210.

See https://github.com/zcash/zips/issues/582 for more details regarding the statement above.

**Comment by conradoplg** [comment]
I managed to recreate the cached state, updated it and it [works](https://github.com/ZcashFoundation/zebra/actions/runs/1713913965) :tada: 

(But now CI is failing :cry: , see https://github.com/ZcashFoundation/zebra/pull/3366)

File: zebra-state/src/service/check/anchors.rs
Comment (dconnolly) [inline zebra-state/src/service/check/anchors.rs]:
```suggestion
 /// Sapling and Orchard do exclusively) _or_ to the interstitial output
```

**Review by dconnolly — DISMISSED** [review]

**Comment by conradoplg** [comment]
@dconnolly I don't know why it dismissed your review, I just accepted your suggestion :confused:

**Review by dconnolly — DISMISSED** [review]

**Comment by conradoplg** [comment]
@dconnolly oops, fixed formatting :sweat_smile:

**Review by dconnolly — APPROVED** [review]

**Pull request merged**
